### PR TITLE
Add JsonKey.required, plus other things

### PR DESCRIPTION
**Pull request by kevmoo**

### CODE REVIEW

File: json_annotation/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,13 @@
+## 0.2.8
+
+* Added `JsonKey.required` field and an associated
+  `MissingRequiredKeysException` that is thrown when `required` fields don't
+  have corresponding keys in a source JSON map.
+
+* Added a new `$` helper function and deprecate an old one. Upgrading to the
```
Comment (jakemac53) [inline json_annotation/CHANGELOG.md]: so mysterious

File: json_annotation/lib/src/allowed_keys_helpers.dart [hunk 2]
```diff
@@ -2,16 +2,38 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
-/// Helper function used in generated code when
+/// **DEPRECATED** Helper function used in generated code when
 /// `JsonSerializable.disallowUnrecognizedKeys` is `true`.
 ///
 /// Should not be used directly.
+@Deprecated('Code generated with the latest `json_serializable` will use '
+    '`\$checkKeys` instead. This function will be removed in the next majoy '
```
Comment (jakemac53) [inline json_annotation/lib/src/allowed_keys_helpers.dart]: s/majoy/major

File: json_annotation/lib/src/allowed_keys_helpers.dart [hunk 3]
```diff
@@ -2,16 +2,38 @@
...
+    'release.')
 void $checkAllowedKeys(Map map, Iterable<String> allowedKeys) {
+  $checkKeys(map, allowedKeys: allowedKeys);
+}
+
+/// Helper function used in generated code when
+/// `JsonSerializable.disallowUnrecognizedKeys` is `true`.
```
Comment (jakemac53) [inline json_annotation/lib/src/allowed_keys_helpers.dart]: Or if any fields have `JsonKey.required`

File: json_annotation/lib/src/allowed_keys_helpers.dart [hunk 4]
```diff
@@ -2,16 +2,38 @@
...
+///
+/// Should not be used directly.
+void $checkKeys(Map map,
+    {Iterable<String> allowedKeys, List<String> requiredKeys}) {
   if (map == null) return;
```
Comment (jakemac53) [inline json_annotation/lib/src/allowed_keys_helpers.dart]: Should this still happen if `requiredKeys` is not empty?

File: json_annotation/lib/src/allowed_keys_helpers.dart [hunk 5]
```diff
@@ -2,16 +2,38 @@
...
+    if (missingKeys.isNotEmpty) {
+      throw new MissingRequiredKeysException(missingKeys, map);
+    }
+  }
+
+  if (allowedKeys != null) {
+    var invalidKeys = map.keys.where((k) => !allowedKeys.contains(k));
+    if (invalidKeys.isNotEmpty) {
+      throw new UnrecognizedKeysException(
+          new List<String>.from(invalidKeys), map, allowedKeys.toList());
```
Comment (jakemac53) [inline json_annotation/lib/src/allowed_keys_helpers.dart]: lets move this `toList` up to the original assignment so its consistent

File: json_serializable/CHANGELOG.md [hunk 6]
```diff
@@ -1,12 +1,18 @@
 ## 0.5.7
 
+* Added support for `JsonKey.required`.
+  * When `true`, generated code throws a `MissingRequiredKeysException` if
+    the key does not exist in the JSON map used to populate the annotated field.
+  * Will be captured captured and wrapped in a `CheckedFromJsonException` if
```
Comment (jakemac53) [inline json_serializable/CHANGELOG.md]: nit: double captured

File: json_serializable/lib/src/generator_helper.dart [hunk 7]
```diff
@@ -128,7 +128,7 @@ class _GeneratorHelper {
         <String, FieldElement>{}, (map, field) {
       if (!field.isPublic) {
         unavailableReasons[field.name] = 'It is assigned to a private field.';
-      } else if (jsonKeyFor(field).ignore == true) {
+      } else if (jsonKeyFor(field).ignore) {
```
Comment (jakemac53) [inline json_serializable/lib/src/generator_helper.dart]: hmm so its important now that anybody using the newest json_serializable changes there json_annotation dependency as well to have the minimum bound of 0.2.8... but there is no way to really communicate that.

I am not sure what the right solution is here?

File: json_annotation/CHANGELOG.md (hunk 1 above)
Comment (kevmoo) [reply json_annotation/CHANGELOG.md]: ack

File: json_annotation/lib/src/allowed_keys_helpers.dart (hunk 2 above)
Comment (kevmoo) [reply json_annotation/lib/src/allowed_keys_helpers.dart]: ack

File: json_annotation/lib/src/allowed_keys_helpers.dart (hunk 3 above)
Comment (kevmoo) [reply json_annotation/lib/src/allowed_keys_helpers.dart]: ack

File: json_annotation/lib/src/allowed_keys_helpers.dart (hunk 5 above)
Comment (kevmoo) [reply json_annotation/lib/src/allowed_keys_helpers.dart]: ack

File: json_serializable/CHANGELOG.md (hunk 6 above)
Comment (kevmoo) [reply json_serializable/CHANGELOG.md]: ack

File: json_serializable/lib/src/generator_helper.dart (hunk 7 above)
Comment (kevmoo) [reply json_serializable/lib/src/generator_helper.dart]: ack

File: json_annotation/lib/src/allowed_keys_helpers.dart (hunk 4 above)
Comment (kevmoo) [reply json_annotation/lib/src/allowed_keys_helpers.dart]: ack – although look at my fix

**Pull request merged**